File: members/U02CDR60PFS/kakao2021/표_편집.cpp
#include <string>
#include <vector>
#include <iostream>
#include <stack>
using namespace std;

struct Node
{
    bool removed;
    Node *prev;
    Node *next;
};

Node nodeArr[1000000];

string solution(int n, int k, vector<string> cmd)
{
    string answer = "";

    for (int i = 1; i < n; i++)
    {
        nodeArr[i - 1].next = &nodeArr[i];
        nodeArr[i].prev = &nodeArr[i - 1];
    }

    Node *cur = &nodeArr[k];
    stack<Node *> stack;

    for (string str : cmd)
    {
        if (str[0] == 'U')
        {
            int x = atoi(str.c_str() + 2);
            for (int i = 0; i < x; i++)
            {
                cur = cur->prev;
            }
        }
        else if (str[0] == 'D')
        {
            int x = atoi(str.c_str() + 2);
            for (int i = 0; i < x; i++)
            {
                cur = cur->next;
            }
        }
        else if (str[0] == 'C')
        {
            stack.push(cur);
            cur->removed = true;
            Node *up = cur->prev;
            Node *down = cur->next;
            if (up)
            {
                up->next = cur->next;
            }
            if (down)
            {
                down->prev = cur->prev;
                cur = down;
            }
            else
            {
                cur = up;
            }
        }
        else
        {
            Node *node = stack.top();
            stack.pop();
            node->removed = false;
            Node *up = node->prev;
            Node *down = node->next;
            if (up)
            {
                up->next = node;
            }
            if (down)
            {
                down->prev = node;
            }
        }
    }

    for (int i = 0; i < n; i++)
    {
        if (nodeArr[i].removed == true)
            answer += 'X';
        else
            answer += 'O';
    }
    return answer;
}

File: members/U02BWCHJA6Q/Programers/kakao_internship_2021/표_편집.js
// https://programmers.co.kr/learn/courses/30/lessons/81303

const JOB = Object.freeze({
  DOWN: 'D',
  UP: 'U',
  CUT: 'C',
  RESTORE: 'Z',
});

function Node(id) {
  this.id = id;
  this.prev = null;
  this.next = null;
}

function LinkedList(n) {
  this.head = null;
  this.current = null;
  this.deletedList = [];

  for (let i = 0; i < n; i++) {
    const node = new Node(i);

    if (this.head === null) {
      this.head = node;
      this.current = node;
    } else {
      this.current.next = node;
      node.prev = this.current;
      this.current = node;
    }
  }

  this.current = this.head;

  this.delete = () => {
    this.deletedList.push(this.current);

    const prev = this.current.prev;
    const next = this.current.next;

    if (prev !== null) {
      prev.next = next;
    } else {
      this.head = next;
    }
    if (next !== null) {
      next.prev = prev;
    }

    next === null ? (this.current = prev) : (this.current = next);
  };

  this.restore = () => {
    const restored = this.deletedList.pop();

    if (restored.prev) {
      restored.prev.next = restored;
    } else {
      this.head = restored;
    }
    if (restored.next) {
      restored.next.prev = restored;
    }
  };
}

function solution(n, k, cmd) {
  const linkedList = new LinkedList(n);

  for (let i = 0; i < k; i++) {
    linkedList.current = linkedList.current.next;
  }

  cmd.forEach(string => {
    const [job, count] = string.split(' ');

    switch (job) {
      case JOB.DOWN:
        {
          for (let i = 0; i < count; i++) {
            linkedList.current = linkedList.current.next;
          }
        }
        break;
      case JOB.UP:
        {
          for (let i = 0; i < count; i++) {
            linkedList.current = linkedList.current.prev;
          }
        }
        break;
      case JOB.CUT:
        {
          linkedList.delete();
        }
        break;
      case JOB.RESTORE: {
        linkedList.restore();
      }
    }
  });

  const answer = Array(n).fill('X');

  linkedList.current = linkedList.head;
  while (linkedList.current) {
    answer[linkedList.current.id] = 'O';
    linkedList.current = linkedList.current.next;
  }

  return answer.join('');
}
